### PR TITLE
feat: add enable stats for peer endpoint of envoyproxy

### LINKED ISSUE #2824: Support Endpoint Stats
*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

Add support for enabling per endpoint stats in envoy 
https://github.com/envoyproxy/envoy/pull/29709

https://gateway.envoyproxy.io/latest/api/extension_types/#proxymetrics

Will help with traffic splitting telemetry https://github.com/envoyproxy/gateway/issues/2809

### PR DESCRIPTION
**Pull request by ShyunnY**
**What type of PR is this?**

feat: add enable stats for peer endpoint of envoyproxy

**Which issue(s) this PR fixes**:

Fixes #2824

### CODE REVIEW

**Comment by ShyunnY** [comment]
Regarding the error in `docs-lint`, is it caused by my modification? I looked at the error message and it appears that I didn't modify the link to the document. Is this a mistake?

**Comment by zirain** [comment]
> Regarding the error in `docs-lint`, is it caused by my modification? I looked at the error message and it appears that I didn't modify the link to the document. Is this a mistake?

please wait https://github.com/envoyproxy/gateway/pull/3146/files

**Comment by ShyunnY** [comment]
hey @arkodg :

I think we should start reviewing this PR as soon as possible, because this PR involves too much xds translator code, and the existing code is constantly changing... This will cause conflicts in this PR.

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 1]
```diff
@@ -29,6 +29,9 @@ type ProxyMetrics struct {
 
 	// EnableVirtualHostStats enables envoy stat metrics for virtual hosts.
 	EnableVirtualHostStats bool `json:"enableVirtualHostStats,omitempty"`
+
+	// EnablePerEndpointStats enables envoy stats metrics for peer endpoint
+	EnablePerEndpointStats bool `json:"enablePerEndpointStats,omitempty"`
```
Comment (zirain) [inline api/v1alpha1/envoyproxy_metric_types.go:34]: In kubernetes world, endpoints may change frequently, would that cause high memory usage?
If so, I suggest EG should highlight that case.

Comment (ShyunnY) [reply api/v1alpha1/envoyproxy_metric_types.go:34]: When the upstream cluster reaches a certain order of magnitude, this "button" becomes a black box. We may need to add comments to the field to inform users to be cautious about enabling this feature.

Comment (zirain) [reply api/v1alpha1/envoyproxy_metric_types.go:34]: yeah, EG definitely should warn/highlight it.

Comment (ShyunnY) [reply api/v1alpha1/envoyproxy_metric_types.go]: done

**Comment by ShyunnY** [comment]
hello, @arkodg 

It looks like this PR has been pending for a long time. Is there anything else I need to do now? I think I should listen to your opinion before trying to resolve the conflict, it could avoid more trouble

**Comment by arkodg** [comment]
hey @ShyunnY will review this one by tomorrow, sorry for the delay on this one

**Comment by ShyunnY** [comment]
> hey @ShyunnY will review this one by tomorrow, sorry for the delay on this one

it does not matter.

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 2]
```diff
@@ -29,6 +29,12 @@ type ProxyMetrics struct {
 
 	// EnableVirtualHostStats enables envoy stat metrics for virtual hosts.
 	EnableVirtualHostStats bool `json:"enableVirtualHostStats,omitempty"`
+
+	// EnablePerEndpointStats enables envoy stats metrics for peer endpoint
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_metric_types.go:33]:
```suggestion
	// EnablePerEndpointStats enables per endpoint envoy stats metrics.
```

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 3]
```diff
@@ -29,6 +29,12 @@ type ProxyMetrics struct {
...
+	// WARNING: This field will let Envoy record the statistics of each endpoint.
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_metric_types.go:34]: this rm L34 and L35

**Comment by arkodg** [comment]
PR looks good ! added a comment around the doc string

File: api/v1alpha1/envoyproxy_metric_types.go (hunk 3 above)
Comment (ShyunnY) [reply api/v1alpha1/envoyproxy_metric_types.go]: a small question, why don’t we keep the warning information?

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**